README.md
Passerelle connector with BAEC services 
=======================================

This passerelle is meant to be connected to BAEC using iMio API Microservices 
(APIMS). APIMS interrogates BOSA's services.

Installation
------------

 - add to Passerelle installed apps settings:
   INSTALLED_APPS += ('passerelle_imio_apims_baec',)

 - enable module:
   PASSERELLE_APP_PASSERELLE_IMIO_APIMS_BAEC_ENABLED = True


Usage
-----

 - create and configure new connector
   - Title/description: whatever you want

 - test service by clicking on the available links
   - the /test/ endpoint to test the connection with APIMS


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [0.0.1] - 2023-04-18
### Created
- Init connector [nselva][jmodesto]

passerelle_imio_apims_casier_judiciaire/models.py
from django.db import models
from django.http import HttpResponse
from django.urls import reverse
from django.core.exceptions import ValidationError
from passerelle.base.models import BaseResource
from passerelle.utils.api import endpoint
from passerelle.utils.jsonresponse import APIError
import requests
from requests import RequestException
import base64
import json
import re

def validate_url(value):
    if value.endswith("/"):
        raise ValidationError(
            '%(value)s ne dois pas finir avec un "/"',
            params={'value': value},
        )

class ApimsCasierJudiciaireConnector(BaseResource):
    """
    Connecteur APIMS Casier Judiciaire
    Attributes
    ----------
    url : str
        url used to connect to APIMS
    username : str
        username used to connect to APIMS
    password : str
        password used to connect to APIMS
    municipality_nis_code : str
        token used to identify municipality to APIMS 
    Methods
    -------
    """
    url = models.URLField(
        max_length=128,
        blank=True,
        verbose_name="URL",
        help_text="URL de APIMS Casier Judiciaire",
        validators=[validate_url],
        default="https://api-staging.imio.be/bosa/v1"
    )
    username = models.CharField(
        max_length=128,
        blank=True,
        help_text="Utilisateur APIMS Casier Judiciaire",
        verbose_name="Utilisateur",
    )
    password = models.CharField(
        max_length=128,
        blank=True,
        help_text="Mot de passe APIMS Casier Judiciaire",
        verbose_name="Mot de passe",
    )
    municipality_nis_code = models.CharField(
        max_length=128,
        blank=True,
        help_text="Code NIS d'identification de l'organisme dans APIMS Casier Judiciaire",
        verbose_name="Code NIS de l'organisme",
    )
    
    category = 'Connecteurs iMio'

    api_description = "Connecteur permettant d'intéragir avec APIMS Casier Judiciaire"

    class Meta:
        verbose_name = 'Connecteur APIMS Casier Judiciaire'
    
    @property
    def session(self):
        session = requests.Session()
        session.auth = (self.username, self.password)
        session.headers.update({
            "Accept": "application/json",
            "X-IMIO-MUNICIPALITY-NIS": self.municipality_nis_code
        })
        return session


    @endpoint(
        name="list-extract-types",
        perm="can_access",
        description="Types d'extraits de casier judiciaire",
        long_description="Lister les différents types d'extraits de casier judiciaire",
        serializer_type="json-api",
        display_order=0,
        parameters={
            "language": {
                "description": "Langage voulu",
                "example_value": "fr",
            },
        },
        display_category="Types",
    )
    def list_person_documents(self, request, language="fr"):
        """ Gets types of extracts
        Returns
        -------
        dict
            all types with reference
        """
        url = f"{self.url}/cjcs-extract-types"

        self.logger.info("Liste des extraits")
        try:
            response = self.session.get(url, params={"language": language})
        except RequestException as e:
            self.logger.warning(f'Casier Judiciaire APIMS Error: {e}')
            raise APIError(f'Casier Judiciaire APIMS Error: {e}')

        json_response = None
        try:
            json_response = response.json()
        except ValueError:
            self.logger.warning('Casier Judiciaire APIMS Error: bad JSON response')
            raise APIError('Casier Judiciaire APIMS Error: bad JSON response')

        try:
            response.raise_for_status()
        except RequestException as e:
            self.logger.warning(f'Casier Judiciaire APIMS Error: {e} {json_response}')
            raise APIError(f'Casier Judiciaire APIMS Error: {e} {json_response}')
        json_response["items"] = sorted(json_response["items"], key=lambda x: int(re.sub(r"\D", "", re.sub(r"(5A)|(5B)", "05",x["code"]))), reverse=False)
        return json_response
    
    @endpoint(
        name="check-extract",
        perm="can_access",
        methods=["get"],
        description="Vérifier la disponibilité du casier judiciaire",
        parameters={
            "extract_code": {
                "description": "ID du type d'extrait de casier judiciaire",
                "example_value": "595",
            },
            "person_nrn": {
                "description": "Numéro de registre national de la personne qui est concernée par l'extrait de casier judiciaire",
                "example_value": "15010123487",
            },
            "requestor_nrn": {
                "description": "Numéro de registre national de la personne qui demande l'extrait de casier judiciaire",
                "example_value": "15010123487",
            },
            "language": {
                "description": "Langage de l'extrait de casier judiciaire",
                "example_value": "fr",
            },
        },
        display_order=1,
        display_category="Documents"
    )
    def check_extract(self, request, extract_code, person_nrn, requestor_nrn, commune_nis=None, language="fr"):
        """ Get asked disponibility document
        Parameters
        ----------
        extract_code : str
            Extract's code
        person_nrn : str
            National number for the extract person
        requestor_nrn : str
            National number of the requester
        language : str
            Language of the document
        Returns
        -------
        JSON
        """
        if commune_nis is None:
            commune_nis = self.municipality_nis_code

        url = f"{self.url}/cjcs-extracts/{person_nrn}/{extract_code}"

        self.logger.info("Récupération du JSON")
        try:
            response = requests.get(
                url,
                auth=(self.username, self.password), 
                headers={
                    "X-IMIO-REQUESTOR-NRN": requestor_nrn,
                    "X-IMIO-MUNICIPALITY-NIS": commune_nis
                },
                params={"language": language}
            )
        except Exception as e:
            self.logger.warning(f'Casier Judiciaire APIMS Error: {e}')
            raise APIError(f'Casier Judiciaire APIMS Error: {e}')

        json_response = None
        try:
            json_response = response.json()
        except ValueError:
            self.logger.warning('Casier Judiciaire APIMS Error: bad JSON response')
            raise APIError('Casier Judiciaire APIMS Error: bad JSON response')

        # try:
        #     response.raise_for_status()
        # except RequestException as e:
        #     self.logger.warning(f'Casier Judiciaire APIMS Error: {e} {json_response}')
        #     raise APIError(f'Casier Judiciaire APIMS Error: {e} {json_response}')
        return json_response
    
    @endpoint(
        name="read-extract",
        perm="can_access",
        methods=["post"],
        description="Lire le casier judiciaire d'une personne",
        display_order=1,
        display_category="Documents"
    )
    def read_extract_demo(self, request):
        """ Get asked document as PDF
        Returns
        -------
        PDF document
        """

        self.logger.info("Casier Judiciaire decode pdf base64")
        body = json.loads(request.body)
        pdf_base64 = body["pdf_base64"]

        pdf_response = None
        try:
            pdf = base64.b64decode(pdf_base64) 
            pdf_response = HttpResponse(pdf, content_type="application/pdf")
        except ValueError:
            self.logger.warning('Casier Judiciaire APIMS Error: bad PDF response')
            raise APIError('Casier Judiciaire APIMS Error: bad PDF response')

        return pdf_response



passerelle_imio_apims_casier_judiciaire/migrations/0002_auto_20230417_1611.py
# Generated by Django 3.2.18 on 2023-04-17 14:11

from django.db import migrations, models
import passerelle_imio_apims_casier_judiciaire.models


class Migration(migrations.Migration):

    dependencies = [
        ('passerelle_imio_apims_casier_judiciaire', '0001_initial'),
    ]

    operations = [
        migrations.AddField(
            model_name='apimscasierjudiciaireconnector',
            name='municipality_nis_code',
            field=models.CharField(blank=True, help_text="Token d'identification de l'organisme dans APIMS Casier Judiciaire", max_length=128, verbose_name="Token d'identification"),
        ),
        migrations.AddField(
            model_name='apimscasierjudiciaireconnector',
            name='password',
            field=models.CharField(blank=True, help_text='Mot de passe APIMS Casier Judiciaire', max_length=128, verbose_name='Mot de passe'),
        ),
        migrations.AddField(
            model_name='apimscasierjudiciaireconnector',
            name='url',
            field=models.URLField(blank=True, default='https://api-staging.imio.be/bosa/v1', help_text='URL de APIMS Casier Judiciaire', max_length=128, validators=[passerelle_imio_apims_casier_judiciaire.models.validate_url], verbose_name='URL'),
        ),
        migrations.AddField(
            model_name='apimscasierjudiciaireconnector',
            name='username',
            field=models.CharField(blank=True, help_text='Utilisateur APIMS Casier Judiciaire', max_length=128, verbose_name='Utilisateur'),
        ),
    ]


passerelle_imio_apims_casier_judiciaire/migrations/0001_initial.py
# Generated by Django 3.2.18 on 2023-04-17 13:01

from django.db import migrations, models


class Migration(migrations.Migration):

    initial = True

    dependencies = [
        ('base', '0030_resourcelog_base_resour_appname_298cbc_idx'),
    ]

    operations = [
        migrations.CreateModel(
            name='ApimsCasierJudiciaireConnector',
            fields=[
                ('id', models.AutoField(auto_created=True, primary_key=True, serialize=False, verbose_name='ID')),
                ('title', models.CharField(max_length=50, verbose_name='Title')),
                ('slug', models.SlugField(unique=True, verbose_name='Identifier')),
                ('description', models.TextField(verbose_name='Description')),
                ('users', models.ManyToManyField(blank=True, related_name='_passerelle_imio_apims_casier_judiciaire_apimscasierjudiciaireconnector_users_+', related_query_name='+', to='base.ApiUser')),
            ],
            options={
                'verbose_name': 'Connecteur APIMS Casier Judiciaire',
            },
        ),
    ]
